presto-main/src/test/java/com/facebook/presto/execution/executor/SimulationTask.java

```diff
@@ -0,0 +1,170 @@
+import io.airlift.units.Duration;
+
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+abstract class SimulationTask
+{
+    private final TaskSpecification specification;
+    private final TaskId taskId;
+
+    private final Set<SimulationSplit> runningSplits = Sets.newConcurrentHashSet();
+    private final Set<SimulationSplit> completedSplits = Sets.newConcurrentHashSet();
+
+    private final TaskHandle taskHandle;
+    private final AtomicBoolean killed = new AtomicBoolean();
+
+    public SimulationTask(TaskExecutor taskExecutor, TaskSpecification specification, TaskId taskId)
+    {
+        this.specification = specification;
+        this.taskId = taskId;
+        taskHandle = taskExecutor.addTask(taskId, () -> 0, 10, new Duration(1, SECONDS));
+    }
+
+    public void setKilled()
+    {
+        killed.set(true);
+    }
+
+    public boolean isKilled()
+    {
+        return killed.get();
+    }
+
+    public Set<SimulationSplit> getCompletedSplits()
+    {
+        return completedSplits;
+    }
+
+    TaskId getTaskId()
+    {
+        return taskId;
+    }
+
+    public TaskHandle getTaskHandle()
+    {
+        return taskHandle;
+    }
+
+    public Set<SimulationSplit> getRunningSplits()
+    {
+        return runningSplits;
+    }
+
+    public synchronized void splitComplete(SimulationSplit split)
+    {
+        runningSplits.remove(split);
+        completedSplits.add(split);
+    }
+
+    public TaskSpecification getSpecification()
+    {
+        return specification;
+    }
+
+    public long getTotalWaitTimeNanos()
+    {
+        long runningWaitTime = runningSplits.stream()
+                .mapToLong(SimulationSplit::getWaitNanos)
+                .sum();
+
+        long completedWaitTime = completedSplits.stream()
+                .mapToLong(SimulationSplit::getWaitNanos)
+                .sum();
+
+        return runningWaitTime + completedWaitTime;
+    }
+
+    public long getProcessedTimeNanos()
+    {
+        long runningProcessedTime = runningSplits.stream()
+                .mapToLong(SimulationSplit::getCompletedProcessNanos)
+                .sum();
+
+        long completedProcessedTime = completedSplits.stream()
+                .mapToLong(SimulationSplit::getCompletedProcessNanos)
+                .sum();
+
+        return runningProcessedTime + completedProcessedTime;
+    }
+
+    public long getScheduledTimeNanos()
+    {
+        long runningWallTime = runningSplits.stream()
+                .mapToLong(SimulationSplit::getScheduledTimeNanos)
+                .sum();
+
+        long completedWallTime = completedSplits.stream()
+                .mapToLong(SimulationSplit::getScheduledTimeNanos)
+                .sum();
+
+        return runningWallTime + completedWallTime;
+    }
+
+    public abstract void schedule(TaskExecutor taskExecutor, int numSplits);
+
+    public static class LeafTask
```

Comment:
You could consolidate these two classes into one that takes `Supplier<SplitSpecification>` and `boolean intermediate`.  The downside is you loose the nice class names that might help with debugging, so just an idea.